src/util/permissions.ts

```diff
@@ -1,256 +1,285 @@
+  if (!channel.guildID) return "8";
+  // Member already has ADMINISTRATOR permission so we return that
+  if (
+    await computeBasePermissions(memberID, channel.guildID) === "8"
+  ) {
+    return "8";
+  }
 
-  // Check if the bot is the owner of the guild, if it is, returns true
-  if (guild.ownerID === botID) return true;
+  const member = await cacheHandlers.get("members", memberID);
+  if (!member) throw Error(Errors.MEMBER_NOT_FOUND);
 
-  const member = (await cacheHandlers.get("members", botID))?.guilds.get(
-    guildID,
+  let permissions = BigInt(0);
+  // First compute @everyone overwrites since these have the lowest priority
+  const overwriteEveryone = channel?.permissionOverwrites.find((overwrite) =>
+    overwrite.id === channel.guildID
   );
-  if (!member) return false;
-
-  // The everyone role is not in member.roles
-  const permissionBits = [...member.roles, guild.id]
-    .map((id) => guild.roles.get(id)!)
-    // Remove any edge case undefined
-    .filter((r) => r)
-    .reduce((bits, data) => {
-      bits |= BigInt(data.permissions);
+  if (overwriteEveryone) {
+    permissions &= ~BigInt(overwriteEveryone.deny);
+    permissions |= BigInt(overwriteEveryone.allow);
+  }
 
-      return bits;
-    }, BigInt(0));
+  const overwrites = channel?.permissionOverwrites;
+
+  let allow = BigInt(0);
+  let deny = BigInt(0);
+  let overwriteRole;
+  // Second compute members role overwrites since these have middle priority
+  for (const roleID in member.guilds.get(channel.guildID)?.roles) {
+    overwriteRole = overwrites.find((overwrite) => overwrite.id === roleID);
+    if (overwriteRole) {
+      deny &= ~BigInt(overwriteRole.deny);
+      allow |= BigInt(overwriteRole.allow);
+    }
+  }
+  // After role overwrite compute save allowed permissions ferst we remove denied permissions since "denied < allowed"
+  permissions &= ~deny;
+  permissions |= allow;
+  // Third compute member specific overwrites since these have the highest priority
+  const overwriteMember = overwrites.find((overwrite) =>
+    overwrite.id === memberID
+  );
+  if (overwriteMember) {
+    permissions &= ~BigInt(overwriteMember.deny);
+    permissions |= BigInt(overwriteMember.allow);
+  }
 
-  if (permissionBits & BigInt(Permissions.ADMINISTRATOR)) return true;
+  return permissions.toString();
+}
 
+/** Checks if the given permissionBits are matching the given Permission[] */
+export function validatePermissions(
+  permissionBits: string,
+  permissions: Permission[],
+) {
   return permissions.every((permission) =>
-    permissionBits & BigInt(Permissions[permission])
+    // Check if permission is in permissionBits
+    BigInt(permissionBits) & BigInt(Permissions[permission])
   );
 }
 
-/** Checks if the bot has the permissions in a channel */
-export function botHasChannelPermissions(
-  channelID: string,
+/** Checks if the given member has these permissions in the given guild */
+export async function hasGuildPermissions(
```

Comment:
`hasChannelPermissions` or `hasRolePermissions` to be more precise.

Reply:
Thats less precise than now.